File: src/nautilus_ctp_adapter/runtime/trading.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum

from .models import CtpRuntimeCommand, CtpRuntimeCommandKind, CtpRuntimeEvent, CtpRuntimeEventKind


class CtpOrderState(StrEnum):
    UNKNOWN = "unknown"
    PENDING_SUBMIT = "pending_submit"
    WORKING = "working"
    PENDING_CANCEL = "pending_cancel"
    PENDING_REPLACE = "pending_replace"
    FILLED = "filled"
    CANCELLED = "cancelled"


@dataclass(slots=True)
class CtpTradingRuntime:
    _order_states: dict[str, CtpOrderState] = field(default_factory=dict)

    def on_command(self, command: CtpRuntimeCommand) -> None:
        client_order_id = command.client_order_id
        if not client_order_id:
            return

        if command.kind is CtpRuntimeCommandKind.SUBMIT_ORDER:
            self._order_states[client_order_id] = CtpOrderState.PENDING_SUBMIT
        elif command.kind is CtpRuntimeCommandKind.CANCEL_ORDER:
            self._order_states[client_order_id] = CtpOrderState.PENDING_CANCEL
        elif command.kind is CtpRuntimeCommandKind.REPLACE_ORDER:
            self._order_states[client_order_id] = CtpOrderState.PENDING_REPLACE

    def on_event(self, event: CtpRuntimeEvent) -> None:
        client_order_id = event.client_order_id
        if not client_order_id:
            return

        if event.kind is CtpRuntimeEventKind.ORDER:
            self._order_states[client_order_id] = CtpOrderState.WORKING
        elif event.kind is CtpRuntimeEventKind.TRADE:
            self._order_states[client_order_id] = CtpOrderState.FILLED
        elif event.kind is CtpRuntimeEventKind.ERROR:
            self._order_states.setdefault(client_order_id, CtpOrderState.UNKNOWN)

    def state_for(self, client_order_id: str) -> CtpOrderState:
        return self._order_states.get(client_order_id, CtpOrderState.UNKNOWN)

    @property
    def tracked_order_count(self) -> int:
        return len(self._order_states)

File: rust/ctp_runtime_core/src/trading.rs
use std::collections::BTreeMap;

use crate::commands::{CtpCommand, CtpCommandKind};
use crate::events::{CtpEvent, CtpEventKind};

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub enum CtpOrderState {
    #[default]
    Unknown,
    PendingSubmit,
    Working,
    PendingCancel,
    PendingReplace,
    Filled,
    Cancelled,
}

#[derive(Debug, Clone, Default)]
pub struct CtpTradingRuntime {
    order_states: BTreeMap<String, CtpOrderState>,
}

impl CtpTradingRuntime {
    pub fn on_command(&mut self, command: &CtpCommand) {
        let Some(client_order_id) = command.client_order_id.as_ref() else {
            return;
        };

        match command.kind {
            CtpCommandKind::SubmitOrder => {
                self.order_states
                    .insert(client_order_id.clone(), CtpOrderState::PendingSubmit);
            }
            CtpCommandKind::CancelOrder => {
                self.order_states
                    .insert(client_order_id.clone(), CtpOrderState::PendingCancel);
            }
            CtpCommandKind::ReplaceOrder => {
                self.order_states
                    .insert(client_order_id.clone(), CtpOrderState::PendingReplace);
            }
            _ => {}
        }
    }

    pub fn on_event(&mut self, event: &CtpEvent) {
        let Some(client_order_id) = event.client_order_id.as_ref() else {
            return;
        };

        match event.kind {
            CtpEventKind::Order => {
                self.order_states
                    .insert(client_order_id.clone(), CtpOrderState::Working);
            }
            CtpEventKind::Trade => {
                self.order_states
                    .insert(client_order_id.clone(), CtpOrderState::Filled);
            }
            CtpEventKind::Error => {
                self.order_states
                    .entry(client_order_id.clone())
                    .or_insert(CtpOrderState::Unknown);
            }
            _ => {}
        }
    }

    pub fn state_for(&self, client_order_id: &str) -> CtpOrderState {
        self.order_states
            .get(client_order_id)
            .copied()
            .unwrap_or(CtpOrderState::Unknown)
    }

    pub fn tracked_order_count(&self) -> usize {
        self.order_states.len()
    }
}
